### ISSUE TITLE
PGD implementation in Jax

### ISSUE THREAD

**Issue by dlanzieri** [DifferentiableUniverseInitiative/jaxpm-paper#2]
Hello @EiffL, I've worked on the implementation of the PGD scheme in JAX. 
In this [notebook ](https://github.com/DifferentiableUniverseInitiative/jaxpm-paper/blob/fcfe7f86d1a0cbb5ad2ae82b3803be38ac40d41f/notebooks/CAMELS_Fitting_pgd.ipynb) I computed the PGD parameters minimizing the loss function over the power spectrum. 
Here are the power spectra obtained using PGD and NN models:
![image](https://user-images.githubusercontent.com/72620117/168598413-20ac1ad7-cf94-4bb6-9c62-1250ec750df2.png)
and the residuals:
![image](https://user-images.githubusercontent.com/72620117/168599008-58524467-0d2f-46e7-be34-5de0ae7e3219.png)
Then, I applied these correction parameters to the results of simulations with different resolutions. You can find the notebook [here](https://github.com/DifferentiableUniverseInitiative/jaxpm-paper/blob/fcfe7f86d1a0cbb5ad2ae82b3803be38ac40d41f/notebooks/NNvsPGDdiffres.ipynb). 
Here are the power spectra using PGD and NN models:
![image](https://user-images.githubusercontent.com/72620117/168599763-5d6c9bcd-2afa-4f0b-b878-baab185b6f8a.png)
the residuals:
![image](https://user-images.githubusercontent.com/72620117/168599865-02594446-850e-4919-954a-e89e47e2287a.png)
And finally, I applied the same correction parameters to the result of simulations with a different resolution, box size, and cosmology. 
You can find the [notebook](https://github.com/DifferentiableUniverseInitiative/jaxpm-paper/blob/fcfe7f86d1a0cbb5ad2ae82b3803be38ac40d41f/notebooks/Simulation-WrongCosmo.ipynb) and see the result plots:
![image](https://user-images.githubusercontent.com/72620117/168600539-6e72d362-61b9-4692-8477-2e5a0f1ed73a.png)
![image](https://user-images.githubusercontent.com/72620117/168600719-746d4a3c-c8bd-46e7-a44f-0ad9612f54a7.png)

**Comment by EiffL** [comment]
This is fantastic @dlanzieri ! That's a lot to unpack here ^^' but it looks great!

I have a few comments  below, but couldl you open separate issues to discuss the  different cosmology and different resolution questions. In this issue we can continue to discuss how to train PGD and its implementation.

#### PGD implementation
- Could you open a PR https://github.com/DifferentiableUniverseInitiative/JaxPM to add the code you added for PGD? 

- ` forces_pgd = forces_pgd * 1.5 * cosmo.Omega_m` why do you apply this factor here? is it in the orginal PGD? I'm not saying it's a bad idea, but I don't remember it was in PGD, which could make it more sensitive to changes in Omega_m

-(tiny detail) In your loss function for training, you don't need to run the odeint ;-) you can run the ode once in this cell
```python
# Precomputing a few data stuff
ref_pos = jnp.stack(poss, axis=0)
ref_vel = jnp.stack(vels, axis=0)
ref_pk = jax.vmap(lambda x: power_spectrum(
      (cic_paint(jnp.zeros(mesh_shape), x)),
      boxsize=np.array([25.] * 3),
      kmin=np.pi / 25.,
      dk=2 * np.pi / 25.)[1])(ref_pos)
```

- Could you make a plot comparing the shape of the PGD kernel and the neural spline correction ? Also, a plot of the field at differentes resolutions

### Different resolution

So, your plot is very interesting. It means that a PGD correction in grid coordinates doesn't translate to different box sizes. One question though, PGD was typically defined in physical scales. 

- Could you try to use the same physical scale for ks, kl in your PGD correction when you go to 128 ?

- In your last notebook, you reuse the PGD params from the 25 Mpc box but at diffeentn cosmo I think... could you check at same cosmo. Just to see the impact of res.

### Different cosmology

It's a bit hard to tell in this last notebook what is going on, because you have different resolutions and differerent cosmo.

- Have a look at my data folder, I think I also downloaded some CAMELS runs at different cosmos. You could use this to look at the effect of different cosmos indepeendently of a change of resolution.

- Can you try for different s8, in addition to different Omega_m? To see what happens for these 2 things.

**Comment by dlanzieri** [comment]
Thanks for the comments @EiffL ! :-)
Right, I'll try to reply section by section! 
 ### PGD implementation
 I've just checked and in fact, in the original PGD, we don't have this factor. The reason why I added it, it's because I started to implement the PGD correction in JAX-PM from scratch, trying to write a force for the short-range interactions looking at the long-range, and in both  FLOWPM and JAX-PM codes these forces were normalized by this factor. 
It's true that being a normalized factor, this will be included in the alpha parameter, but as you said, this means we have a different sensitivity to changes in Omega_m compared to the original PGD.

**Comment by dlanzieri** [comment]
### PGD implementation
@EiffL Concerning the PR for JAX-PM, I haven't been able to push my modifications to the repository, I don’t know if I didn't clone the repository well or if you have just to allow me to contribute.

### Different cosmology
In this [notebook](https://github.com/DifferentiableUniverseInitiative/jaxpm-paper/blob/a9d03a4b8779a00734da53ad6587ac52df3de624/notebooks/NNvsPGD_wrong_cosmo_omega_M.ipynb), I used the Camels simulations with a different Omega_m, so the results we see are independent of a change of resolution. 
Here we can see some plots:
<img width="575" alt="Schermata 2022-05-17 alle 17 24 24" src="https://user-images.githubusercontent.com/72620117/168848622-0425a78f-a49d-4796-adc3-3cf27064d1d3.png">
<img width="574" alt="Schermata 2022-05-17 alle 17 24 41" src="https://user-images.githubusercontent.com/72620117/168848630-6507e210-6880-4b90-af8e-2feb56e16852.png">

As you suggested, could be interested to try different sigma8,  however in your data folder I just found 4 sets of sims, 3 of those have the same Omega_m and sigma8 we used to train the correction schemes, the other one has a different Omega_m (it's the set I used in the notebook above ). If I could download the simulation with a different sigma8, we would be able to make a fair comparison.  Is the set of simulations accessible to everyone?

**Comment by EiffL** [comment]
I've just given you writing rights on the jaxpm repo @dlanzieri. Great plot :-) I'll download some other sims. They are all available from the CAMELS website

**Comment by dlanzieri** [comment]
## Different cosmology - PGD implementation

**Comment by dlanzieri** [comment]
##  Different cosmology- PGD implementation
I'm remaking the plots with the new PGD-parameters (those found without the factor` forces_pgd = forces_pgd * 1.5 * cosmo.Omega_m`). I thought it was interesting to document the results when those are applied to the cosmology with a different Omega_m.
With the scaling with omega_m:
<img width="1183" alt="Schermata 2022-05-18 alle 11 27 29" src="https://user-images.githubusercontent.com/72620117/169006736-904f6281-d714-4446-a4eb-549d048cbada.png">
Without the scaling with omega_m:
<img width="1185" alt="Schermata 2022-05-18 alle 11 27 41" src="https://user-images.githubusercontent.com/72620117/169006747-4f2f3033-ae9e-4e9c-82ca-92d708cc4ef1.png">
